src/coreclr/src/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs

```diff
@@ -486,50 +486,209 @@ private static int CalculateOptimalMaxSpinWaitsPerSpinIteration()
+            // TODO: Consider flushing the currentProcessorIdCache on Wait operations or similar
+            // actions that are likely to result in changing the executing core
+            [ThreadStatic]
+            private static int t_currentProcessorIdCache;
+
+            private const int ProcessorIdCacheShift = 16;
+            private const int ProcessorIdCacheCountDownMask = (1 << ProcessorIdCacheShift) - 1;
+            // 50 is our best guess.
+            // Based on further calibration it is likley to be adjusted lower.
+            // In relatively rare cases of a slow GetCurrentProcessorNumber, it may be recalibrated to a higher number.
+            private static int s_processorIdRefreshRate = 50;
+            // We will not adjust higher than this though.
+            private const int MaxIdRefreshRate = 5000;
+
+            private static int RefreshCurrentProcessorId()
+            {
+                double[]? calibrationSamples = s_CalibrationSamples;
+                if (calibrationSamples != null)
+                    CalibrateOnce(calibrationSamples);
+
+                int currentProcessorId = GetCurrentProcessorNumber();
+
+                // On Unix, GetCurrentProcessorNumber() is implemented in terms of sched_getcpu, which
+                // doesn't exist on all platforms.  On those it doesn't exist on, GetCurrentProcessorNumber()
+                // returns -1.  As a fallback in that case and to spread the threads across the buckets
+                // by default, we use the current managed thread ID as a proxy.
+                if (currentProcessorId < 0) currentProcessorId = Environment.CurrentManagedThreadId;
 
-            // On Unix, GetCurrentProcessorNumber() is implemented in terms of sched_getcpu, which
-            // doesn't exist on all platforms.  On those it doesn't exist on, GetCurrentProcessorNumber()
-            // returns -1.  As a fallback in that case and to spread the threads across the buckets
-            // by default, we use the current managed thread ID as a proxy.
-            if (currentProcessorId < 0) currentProcessorId = Environment.CurrentManagedThreadId;
+                Debug.Assert(s_processorIdRefreshRate <= ProcessorIdCacheCountDownMask);
 
-            // Add offset to make it clear that it is not guaranteed to be 0-based processor number
-            currentProcessorId += 100;
+                // Mask with int.MaxValue to ensure the execution Id is not negative
+                t_currentProcessorIdCache = ((currentProcessorId << ProcessorIdCacheShift) & int.MaxValue) | s_processorIdRefreshRate;
+
+                return currentProcessorId;
+            }
 
-            Debug.Assert(ProcessorIdRefreshRate <= ProcessorIdCacheCountDownMask);
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            internal static int GetCurrentProcessorId()
+            {
+                if (s_processorIdRefreshRate <= 2)
```

Comment:
This is aggressively inlined method, so this will get stamped at every callsite. It would be really nice to make `SimpleCoreIdSpeedCheck` reliable-enough (see my other feedback) to make this unnecessary.

Reply:
That was added because I have seen machines that are exactly on 2x "fence". My main machine included. So sometimes they pass the first test sometimes not.

Now after testing this on various environments I see that access has a bit of clustering nature likely indicating changes in underlying implementation:
- < 0.5 of TLS .  (RDPID or similar solution)
- 1-2x of TLS .    (RDTSP )
- 20x and more  (some syscall indirection)

I think making the criteria in `SimpleCoreIdSpeedCheck` to be 3x will make this condition in `GetCurrentProcessorId` unnecessary.

I need to play with this a bit more on Linux/ARM to see if the pattern is similar.

Reply:
3x criteria for "fast" is indeed more reliable and makes the check in the caching path unnecessary. 

I tried this on a Linux VM in Azure and it behaves in a similar way.
By the timings it looks like RDTSP is used. 

Linux reports 1000000000 frequency (1ns tick) - more likely it is the resolution, cannot be the frequency, but these measurements do not depend on that anyways.
Otherwise Linux seems to have slightly faster `GetProcessorNumber` and slightly slower TLS. - assuming both VMs run on similar hardware. 

== Azure VM Ubuntu (one access in ticks)
ID: 11.52392578125  TLS: 6.982421875

== Azure VM Win10
ID: 0.1416015625  TLS: 0.05615234375


Linux ARM64 machine that I could try this on is "slow" and is not sensitive to this change.

Reply:
Tried this on a newer Windows ARM64 machine. Performance is similar to x64 case with RDPID
I assume this version of ARM64 has hardware support for core ID and OS uses it.

=== on Windows ARM64 (64 logical cores, 2.2 GHz)
`StopWatch` frequency: 200000000
1 tick is 5ns

times (in ticks)
ID: 1.098876953125  TLS: 1.54833984375

On this platform `GetProcessorNumber` is firmly detected as "fast no caching needed".
This measurement includes  #581